compiler/lib/inline.ml

```diff
@@ -26,21 +26,72 @@ type prop =
   ; optimizable : bool
   }
 
+type closure_info =
+  { cl_params : Var.t list
+  ; cl_cont : int * Var.t list
+  ; cl_prop : prop
+  ; cl_simpl : (Var.Set.t * bool * Var.Set.t) option
+  }
+
+let block_size { branch; body; _ } =
+  List.length body
+  +
+  match fst branch with
+  | Cond _ -> 2
+  | Switch (_, a1) -> Array.length a1
+  | _ -> 0
+
+let simple_function blocks name params pc =
+  let bound_vars =
+    ref
+      (List.fold_left
+         ~f:(fun s x -> Var.Set.add x s)
+         ~init:Var.Set.empty
+         (name :: params))
+  in
+  let recursive = ref false in
+  let tc = ref Var.Set.empty in
+  try
+    Code.preorder_traverse
+      { fold = Code.fold_children }
+      (fun pc () ->
+        let block = Addr.Map.find pc blocks in
+        (match block.branch with
+        | Pushtrap _, _ -> raise Exit
+        | Return x, _ -> (
+            match List.last block.body with
+            | None -> ()
+            | Some (Let (y, Apply { f; _ }), _) ->
+                if Code.Var.equal x y && List.mem f ~set:params
+                then tc := Var.Set.add f !tc
+            | Some _ -> ())
+        | _ -> ());
+        List.iter block.body ~f:(fun (i, _loc) ->
+            match i with
+            | Let (_, Closure _) -> raise Exit
+            | _ -> ());
+        Freevars.iter_block_bound_vars
+          (fun x -> bound_vars := Var.Set.add x !bound_vars)
+          block;
+        Freevars.iter_block_free_vars
+          (fun x ->
+            if not (Var.Set.mem x !bound_vars) then raise Exit;
```

Comment:
@vouillon, could you explain why we have this restriction ?

Reply:
To inline a function, we have to make sure that all its free variables are still in scope. This is a simple way to ensure that.

Reply:
Thinking more about it, I fail to see how free variables could be out of scope. Could you elaborate a bit more ?

Removing the restriction would require that we keep live_vars up to date when duplicate free vars usage.

Reply:
Oh, indeed this is probably not necessary since we are only inlining functions that are defined in the same scope or in an outer scope.